File: scripts/executor.py
import argparse
import concurrent.futures
import subprocess
from os import cpu_count, getloadavg
from os.path import exists
import time

from tsprd_data import *

TIME_LIMIT = int(10 * 60 * (1976.0 / 1201.0))


def tsprd_execute(output_folder: str, n_threads: int):
    save_git_commit_hash(output_folder)
    instances = get_instances_execs()

    with concurrent.futures.ThreadPoolExecutor(max_workers=n_threads) as executor:
        total_elements = len(instances)
        current_element = 0
        start_time = time.time()
        print(" Executed  |   Time   |  Load   | Last Instance")
        for iset, name, beta, exec_id in executor.map(lambda i: execute_instance(*i, output_folder), instances):
            total_cores = cpu_count()
            current_element += 1
            current_time = time.time() - start_time
            instance_name = "{}/{}_{} {}".format(iset, name, beta, exec_id)

            print("\r {:4d}/{:4d} | {:02d}:{:02d}:{:02d} | {:2.2f}/{:<2d} | {}".format(current_element, total_elements, int(current_time / 3600), int(
                current_time / 60 % 60), int(current_time % 60), getloadavg()[0], total_cores, instance_name, end="   ", flush=True))


def save_git_commit_hash(output_folder: str):
    subprocess.run(["mkdir -p {}".format(output_folder)], stdout=subprocess.DEVNULL, shell=True)
    git_hash = subprocess.check_output(['git', 'rev-parse', 'HEAD']).decode('ascii').strip()
    print(git_hash, file=open("{}/git-commit.hash".format(output_folder), 'w'))


def execute_instance(iset, _, name, beta, exec_id, output_folder):
    instance = "{}/{}_{}".format(iset, name, beta)
    instance_file = "../instances/{}.dat".format(instance)
    output_file = "{}/{}_{}.txt".format(output_folder, instance, exec_id)

    if not exists(output_file):  # skip if it was already executed and saved
        command = "julia executor.jl {} -o {} -t {}".format(instance_file, output_file, TIME_LIMIT)
        process = subprocess.run([command], stdout=subprocess.DEVNULL, shell=True)

        if process.returncode != 0:
            print(instance, file=open("{}/errors.txt".format(output_folder), 'a'))
            raise Exception("Error while running {}".format(instance))

    return iset, name, beta, exec_id


def main():
    parser = argparse.ArgumentParser(description='Run all instances of TSPrd.')
    parser.add_argument("output_folder", action="store",
                        type=str, help="Which folder to save the output files.")
    parser.add_argument("n_threads", action="store", type=int, nargs="?",
                        default=10, help="Maximum number of threads to execute concurrently.")
    args = parser.parse_args()
    tsprd_execute(args.output_folder, args.n_threads)


if __name__ == "__main__":
    main()

File: scripts/executor.jl
using HgsTSPrd, Printf, Logging

"""
    tsprd_execute(outputfolder, threads, threadid)

Execute all TSPrd instances, saving the results in `outputfolder`
The variables `threads` and `threadid` is used to split the instances between threads.
"""
function tsprd_execute(outputfolder::String, threads::Int, threadid::Int)
    mkpath(outputfolder)
    run(pipeline(`git rev-parse HEAD`; stdout = joinpath(outputfolder, "git.commit"))) # save git commit hash to file

    instanceids = getinstancesids()
    sort!(instanceids; by = iid -> iid.N)
    instanceids = [instanceids[i] for i in threadid:threads:lastindex(instanceids)] # instances for this thread

    println(" Executed  |   Time   | Last Instance")
    currentelement = 0
    startime = time_ns()
    ninstances = length(instanceids)

    for instanceid in instanceids
        runinstance(instanceid, outputfolder)

        instancename = "$(instanceid.instanceset)/$(instanceid.name)_$(instanceid.beta)"
        currenttime = (time_ns() - startime) / 1000000000
        currentelement += 1
        @printf(
            "\r %4d/%4d | %02d:%02d:%02d | %s                  ",
            currentelement,
            ninstances,
            (currenttime ÷ 3600),
            (currenttime ÷ 60 % 60),
            (currenttime % 60),
            instancename
        )
    end

    return nothing
end

const TIME_LIMIT = floor(Int, 10 * 60 * (1976.0 / 1201.0))

struct InstanceID
    instanceset::String
    name::String
    N::Int
    beta::String

    function InstanceID(iset::String, name::String, beta::String)
        N = parse(Int, match(r"\d+", name).match)
        return new(iset, name, N, beta)
    end
end

function runinstance(instanceid::InstanceID, outputfolder::String)
    instance_str = "$(instanceid.instanceset)/$(instanceid.name)_$(instanceid.beta)"
    inputfile = "../instances/$instance_str.dat"
    data = Data(String[inputfile, "-t", string(TIME_LIMIT)])
    ga = GeneticAlgorithm(data)

    execute!(ga) # warmup for compilation

    for execid in 1:10
        outputfile = joinpath(outputfolder, "$(instance_str)_$(execid).txt")
        isfile(outputfile) && continue # already executed
        execute!(ga, outputfile)
    end

    return nothing
end

function getinstancesids()
    #! format: off
    solomon_instances = [("Solomon", "$n/$name") for n in (10, 15, 20, 25, 30, 35, 40, 45, 50, 100) for name in ("C101", "C201", "R101", "RC101")]
    tsplib_names = ["eil51", "berlin52", "st70", "eil76", "pr76", "rat99", "kroA100", "kroB100", "kroC100", "kroD100", "kroE100", "rd100", "eil101",
                    "lin105", "pr107", "pr124", "bier127", "ch130", "pr136", "pr144", "ch150", "kroA150", "kroB150", "pr152", "u159", "rat195",
                    "d198", "kroA200", "kroB200", "ts225", "tsp225", "pr226", "gil262", "pr264", "a280", "pr299", "lin318", "rd400", "fl417", "pr439",
                    "pcb442", "d493"]
    tsplib_instances = [["TSPLIB", name] for name in tsplib_names]
    atsplib_instances = [["aTSPLIB", name] for name in ["ftv33", "ft53", "ftv70", "kro124p", "rbg403"]]
    #! format: on
    all_instances = Iterators.flatten([solomon_instances, atsplib_instances, tsplib_instances])
    betas = ["0.5", "1", "1.5", "2", "2.5", "3"]

    return [InstanceID(instanceset, name, beta) for (instanceset, name) in all_instances for beta in betas]
end

if length(ARGS) != 3
    @warn("Usage: executor.jl output_folder nthreads threadid")
    exit(1)
elseif !isdirpath("$(ARGS[1])/")
    @warn("\"$(ARGS[1])\" is not a valid folder.")
    exit(1)
end

threads = parse(Int, ARGS[2])
threadid = parse(Int, ARGS[3])

if (threadid > threads)
    @warn "Invalid thread id $threadid for $threads threads."
    exit(1)
end

disable_logging(Logging.Info)
tsprd_execute(ARGS[1], threads, threadid)
